### ISSUE TITLE
Allow objects as providers

### ISSUE THREAD

**Issue by jl-wynen** [scipp/sciline#114]
I don't see any a priori reason for only allowing functions as providers. Any callable should work. For example, the following is equivalent to a function with bound parameters (with `functools.partial`):
```python
from typing import NewType
from sciline import Pipeline

Data = NewType('Data', int)

class Provider:
    def __init__(self, param: int) -> None:
        self._param = param

    def __call__(self) -> Data:
        return Data(3 * self._param)

pl = Pipeline((Provider(2),))
```
But it raises
```
TypeError: <__main__.Provider object at 0x7fa79eb4cfd0> is not a module, class, method, or function.
```

Supporting this would

- allow for more flexibility. Such a provider can maintain internal state as long as that state does not affect the output value (e.g. caches). Or it could be used to inject objects that don't exist at the point where `Pipeline.build` gets called.
- bring Sciline closer to the OOP world that is, in my experience, more familiar to most Python programmers than the functional world.

**Comment by SimonHeybrock** [comment]
I think I tried this during initial implementation, but had trouble extracting the type hints. I suppose it is possible somehow.

> Supporting this would
> - allow for more flexibility. Such a provider can maintain internal state as long as that state does not affect the output value (e.g. caches). Or it could be used to inject objects that don't exist at the point where Pipeline.build gets called.

Note that you can already use caching and other bits as it is, this is not limited to classes but can be done with functions.

**Comment by jl-wynen** [comment]
> Interesting. Many libraries used in the scientific world are basically functional (NumPy, Pandas, ...), who do you have in mind here?

Those libraries cannot be customised. Look, e.g., at [scipy.stats](https://docs.scipy.org/doc/scipy/reference/stats.html). In my experience, most people learn OOP for anything that goes beyond calling simple functions. But I also don't want to encourage overuse of classes.

> Note that you can already use caching and other bits as it is, this is not limited to classes but can be done with functions.

True, because functions are objects in Python. But the syntax is less intuitive.

The main reason I opened this issue is that I found it surprising that class instances are not supported because providers are just `Callable[[...], Any]`. So there is no restriction to functions in the interface of Sciline.
